File: agents/dart/test/openlog_test.dart
import 'dart:convert';

import 'package:openlog/openlog.dart';
import 'package:test/test.dart';

class Captured {
  final List<String> bodies = [];
  Future<int> send(String url, Map<String, String> headers, String body) async {
    bodies.add(body);
    return 200;
  }
}

OpenlogOptions opts({double sampleRate = 1, int batch = 1}) => OpenlogOptions(
      key: 'olb_1a2b3c4d5e6f708192a3b4c5d6e7f809',
      endpoint: 'https://ingest.example.com:4318',
      appId: 'com.example.shop',
      sampleRate: sampleRate,
      maxBatchSize: batch,
    );

Future<Map<String, Object?>?> noConfig(String url, Map<String, String> headers) async => null;

List<Map<String, Object?>> spansOf(String body) {
  final decoded = jsonDecode(body) as Map<String, Object?>;
  final rs = (decoded['resourceSpans']! as List<Object?>).first! as Map<String, Object?>;
  final ss = (rs['scopeSpans']! as List<Object?>).first! as Map<String, Object?>;
  return (ss['spans']! as List<Object?>).cast<Map<String, Object?>>();
}

Map<String, String> attrsOf(Map<String, Object?> span) => {
      for (final a in (span['attributes']! as List<Object?>).cast<Map<String, Object?>>())
        a['key']! as String: (a['value']! as Map<String, Object?>)['stringValue']! as String,
    };

void main() {
  tearDown(Openlog.reset);

  test('init twice returns the first instance', () {
    final a = Openlog.init(opts(), send: Captured().send, fetchConfig: noConfig);
    final b = Openlog.init(opts(), send: Captured().send, fetchConfig: noConfig);
    // A second agent would double-count every screen, and the usual cause is a framework starting twice.
    expect(identical(a, b), isTrue);
  });

  test('a screen carries its route and the session', () async {
    final server = Captured();
    final sdk = Openlog.init(opts(), send: server.send, fetchConfig: noConfig);
    sdk.recordScreen('/cart', isColdStart: true);
    await sdk.flush();

    final attrs = attrsOf(spansOf(server.bodies.single).single);
    expect(attrs['openlog.rum.event'], 'page_view');
    expect(attrs['openlog.rum.route'], '/cart');
    expect(attrs['openlog.rum.page_view.kind'], 'load');
    expect(attrs['session.id'], matches(RegExp(r'^[0-9a-f]{32}$')));
  });

  test('identify attaches an id to later events, and clears it', () async {
    final server = Captured();
    final sdk = Openlog.init(opts(batch: 10), send: server.send, fetchConfig: noConfig);

    sdk.recordEvent('before');
    sdk.identify('  acct_8f3a2b  ');
    sdk.recordEvent('after');
    sdk.identify('');
    sdk.recordEvent('signed_out');
    await sdk.flush();

    final spans = spansOf(server.bodies.single);
    final byName = {for (final s in spans) s['name']! as String: attrsOf(s)};
    expect(byName['before']!.containsKey('user.id'), isFalse);
    expect(byName['after']!['user.id'], 'acct_8f3a2b', reason: 'trimmed, and attached to later spans only');
    expect(byName['signed_out']!.containsKey('user.id'), isFalse, reason: 'absent, not empty');
  });

  test('an over-long identity is bounded, as the server bounds it', () async {
    final server = Captured();
    final sdk = Openlog.init(opts(), send: server.send, fetchConfig: noConfig);
    sdk.identify('u' * 500);
    sdk.recordEvent('checkout');
    await sdk.flush();
    // Bounded here as well as on the server: an application should not discover the limit by having its
    // spans silently change shape somewhere it cannot see.
    expect(attrsOf(spansOf(server.bodies.single).single)['user.id']!.length, 128);
  });

  test('custom parameters are namespaced, capped and deterministic', () async {
    final server = Captured();
    final sdk = Openlog.init(opts(), send: server.send, fetchConfig: noConfig);
    sdk.recordEvent('checkout', params: {
      for (var i = 0; i < 25; i++) 'p${i.toString().padLeft(2, '0')}': i,
      'Bad Key': 'dropped',
    });
    await sdk.flush();

    final attrs = attrsOf(spansOf(server.bodies.single).single);
    final params = attrs.keys.where((k) => k.startsWith('openlog.rum.custom.param.')).toList()..sort();
    expect(params.length, 16, reason: 'the cap bounds what one event can carry');
    // Sorted before the cap: which sixteen survive depends on the payload, not on map order.
    expect(params.first, 'openlog.rum.custom.param.p00');
    expect(attrs.containsKey('openlog.rum.custom.param.Bad Key'), isFalse);
  });

  test('a timing carries its value and unit', () async {
    final server = Captured();
    final sdk = Openlog.init(opts(), send: server.send, fetchConfig: noConfig);
    sdk.recordTiming('cart_priced', 42);
    await sdk.flush();
    final attrs = attrsOf(spansOf(server.bodies.single).single);
    expect(attrs['openlog.rum.custom.value'], '42');
    expect(attrs['openlog.rum.custom.unit'], 'ms');
  });

  test('an error carries the exception as a span event', () async {
    final server = Captured();
    final sdk = Openlog.init(opts(), send: server.send, fetchConfig: noConfig);
    sdk.recordError(StateError('cart is empty'), stack: StackTrace.current);
    await sdk.flush();
    final span = spansOf(server.bodies.single).single;
    expect(span['status'], {'code': 2});
    expect(span['events'], isNotEmpty);
  });

  test('a sampled-out session sends nothing at all', () async {
    final server = Captured();
    // Half a session is not a cheaper session, it is an unreadable one.
    final sdk = Openlog.init(opts(sampleRate: 0.0000001), send: server.send, fetchConfig: noConfig);
    if (sdk.sessionId.isEmpty) {
      sdk.recordScreen('/cart');
      sdk.recordEvent('checkout');
      await sdk.flush();
      expect(server.bodies, isEmpty);
    }
  });

  test('the server sample rate replaces the compiled-in one', () async {
    final server = Captured();
    Openlog.init(opts(), send: server.send, fetchConfig: (url, h) async {
      expect(url, endsWith('/v1/rum/config'));
      expect(h['openlog-browser-key'], startsWith('olb_'));
      return {'service_name': 'shop-android', 'environment': 'production', 'sample_rate': 0.5};
    });
    // The fetch is deliberately not awaited by init; letting the microtask queue drain is enough.
    await Future<void>.delayed(Duration.zero);
  });
}

File: agents/kotlin/src/test/kotlin/io/github/onuragtas/openlog/OpenLogTest.kt
package io.github.onuragtas.openlog

import kotlin.test.AfterTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

class OpenLogTest {
    private class Captured {
        val bodies = ArrayList<String>()
        fun send(url: String, h: Map<String, String>, body: String): Int {
            bodies.add(body)
            return 200
        }
    }

    @AfterTest
    fun tearDown() = OpenLog.reset()

    private fun opts(sampleRate: Double = 1.0, batch: Int = 1) = Options(
        key = "olb_1a2b3c4d5e6f708192a3b4c5d6e7f809",
        endpoint = "https://ingest.example.com:4318",
        appId = "com.example.shop",
        sampleRate = sampleRate,
        maxBatchSize = batch,
    )

    private val noConfig: ConfigFetch = { _, _ -> null }

    /**
     * The attributes of the span named [name], read back out of the encoded body.
     *
     * The body is split at span boundaries first. Matching across the whole body with a lazy pattern reads
     * one span's attributes into another's — which is exactly what this helper did at first, and it
     * reported a bug in identify() that was not there.
     */
    private fun attrsOf(body: String, name: String): Map<String, String> {
        val span = body.split("""{"traceId":""").firstOrNull { it.contains(""""name":"$name"""") }
            ?: error("no span named $name in $body")
        return Regex(""""key":"([^"]+)","value":\{"stringValue":"([^"]*)"\}""")
            .findAll(span).associate { it.groupValues[1] to it.groupValues[2] }
    }

    @Test
    fun `start twice returns the first instance`() {
        val a = OpenLog.start(opts(), Captured()::send, noConfig)
        val b = OpenLog.start(opts(), Captured()::send, noConfig)
        // A second SDK would double-count every screen, and the usual cause is a framework starting twice.
        assertTrue(a === b)
    }

    @Test
    fun `a screen carries its route and the session`() {
        val server = Captured()
        val sdk = OpenLog.start(opts(), server::send, noConfig)
        sdk.recordScreen("/cart", isColdStart = true)

        val a = attrsOf(server.bodies[0], "screen /cart")
        assertEquals("page_view", a["openlog.rum.event"])
        assertEquals("/cart", a["openlog.rum.route"])
        assertEquals("load", a["openlog.rum.page_view.kind"])
        assertTrue(Regex("^[0-9a-f]{32}$").matches(a["session.id"]!!))
    }

    @Test
    fun `identify attaches an id to later events and clears it`() {
        val server = Captured()
        val sdk = OpenLog.start(opts(batch = 10), server::send, noConfig)
        sdk.recordEvent("before")
        sdk.identify("  acct_8f3a2b  ")
        sdk.recordEvent("after")
        sdk.identify("")
        sdk.recordEvent("signed_out")
        sdk.flush()

        val body = server.bodies[0]
        assertNull(attrsOf(body, "before")["user.id"])
        assertEquals("acct_8f3a2b", attrsOf(body, "after")["user.id"], "trimmed, and on later spans only")
        assertNull(attrsOf(body, "signed_out")["user.id"], "absent, not empty")
        assertEquals(
            1, Regex(""""key":"user\.id"""").findAll(body).count(),
            "exactly one of the three spans carries an identity",
        )
    }

    @Test
    fun `an over-long identity is bounded`() {
        val server = Captured()
        val sdk = OpenLog.start(opts(), server::send, noConfig)
        sdk.identify("u".repeat(500))
        sdk.recordEvent("checkout")
        // Bounded here as well as on the server: an application should not discover the limit by having
        // its spans silently change shape somewhere it cannot see.
        assertEquals(128, attrsOf(server.bodies[0], "checkout")["user.id"]!!.length)
    }

    @Test
    fun `custom parameters are namespaced capped and deterministic`() {
        val server = Captured()
        val sdk = OpenLog.start(opts(), server::send, noConfig)
        val params = HashMap<String, Any?>()
        for (i in 0 until 25) params[String.format("p%02d", i)] = i
        params["Bad Key"] = "dropped"
        sdk.recordEvent("checkout", params)

        val a = attrsOf(server.bodies[0], "checkout")
        val keys = a.keys.filter { it.startsWith("openlog.rum.custom.param.") }.sorted()
        assertEquals(16, keys.size, "the cap bounds what one event can carry")
        assertEquals("openlog.rum.custom.param.p00", keys.first(), "sorted before the cap, not map order")
        assertNull(a["openlog.rum.custom.param.Bad Key"])
    }

    @Test
    fun `a timing carries its value and unit`() {
        val server = Captured()
        val sdk = OpenLog.start(opts(), server::send, noConfig)
        sdk.recordTiming("cart_priced", 42)
        val a = attrsOf(server.bodies[0], "cart_priced")
        assertEquals("42", a["openlog.rum.custom.value"])
        assertEquals("ms", a["openlog.rum.custom.unit"])
    }

    @Test
    fun `an error carries the exception as a span event`() {
        val server = Captured()
        val sdk = OpenLog.start(opts(), server::send, noConfig)
        sdk.recordError(IllegalStateException("cart is empty"))
        assertTrue(server.bodies[0].contains("exception.type"))
        assertTrue(server.bodies[0].contains(""""status":{"code":2}"""))
    }

    @Test
    fun `the config response is read without a JSON parser`() {
        val got = parseConfigFields("""{"service_name":"shop-android","environment":"production","sample_rate":0.5}""")
        assertEquals("shop-android", got["service_name"])
        assertEquals("production", got["environment"])
        assertEquals(0.5, got["sample_rate"])
        // A field that is absent is simply absent, which is the same as the call having failed.
        assertNull(parseConfigFields("""{"service_name":"x"}""")["sample_rate"])
        assertNull(parseConfigFields("""{"sample_rate":"half"}""")["sample_rate"])
    }
}
